Do not overflow intermediate

erights opened the pull request:
Use BigInt inside `getOutputPrice` so intermediate integers do not overflow.

Also fixes #2193

Chris-Hibbert commented:
As long as you're working here, wanna fix #2193 as well? (I only reported it this morning.)

Chris-Hibbert commented on packages/zoe/src/contractSupport/bondingCurves.js:
```diff
@@ -32,12 +33,12 @@ export const getInputPrice = (
-  const inputWithFee = multiply(inputValue, oneMinusFeeInTenThousandths);
-  const numerator = multiply(inputWithFee, outputReserve);
-  const denominator = add(multiply(inputReserve, 10000), inputWithFee);
+  const oneMinusFeeInTenThousandths = BigInt(10000) - BigInt(feeBasisPoints);
```
Should we pull `BigInt(10000)` out as a manifest constant?  (`TEN_THOUSAND`)

Chris-Hibbert replied:
done

Chris-Hibbert commented on packages/zoe/src/contractSupport/bondingCurves.js:
```diff
@@ -32,12 +33,12 @@ export const getInputPrice = (
-  const inputWithFee = multiply(inputValue, oneMinusFeeInTenThousandths);
-  const numerator = multiply(inputWithFee, outputReserve);
-  const denominator = add(multiply(inputReserve, 10000), inputWithFee);
+  const oneMinusFeeInTenThousandths = BigInt(10000) - BigInt(feeBasisPoints);
```
'inTenThousandths' sounds wrong to me.  Shouldn't it be `oneMinusFeeTimesTenThousand`?

Chris-Hibbert commented on packages/zoe/src/contractSupport/bondingCurves.js:
```diff
@@ -65,14 +68,14 @@ export const getOutputPrice = (
+  const numerator = BigInt(outputValue) * BigInt(inputReserve) * BigInt(10000);
+  const denominator =
+    (BigInt(outputReserve) - BigInt(outputValue)) * oneMinusFeeInTenThousandths;
+  if (denominator <= BigInt(0)) {
```
The conditional on line 40 can only be true if `inputValue` is negative. This check won't trigger if `outputValue` is negative, but smaller than outputReserve. 

I think we need to check if outputValue is negative or if (outputValue > outputReserve).

Chris-Hibbert reviewed (commented):
Please add some tests. The changes look right, but without tests that fail before and succeed after, I wouldn't be sure.

erights commented:
> As long as you're working here, wanna fix #2193 as well? (I only reported it this morning.)

I added code to do as #2193 suggests. But why is 0 the right answer in that case? Isn't the error actually better?

I also added similar code for getInputPrice. Negative seems impossible but I don't know that zero is impossible. But is the divide error in that case better or is returning zero better?

I'll wait for answers on both before merging.

Chris-Hibbert commented:
> I added code to do as #2193 suggests. But why is 0 the right answer in that case? Isn't the error actually better?

In the UI, the error is a problem, and it was caused when the user (me) entered a number that was too large. I don't think the UI can detect that easily, but probably something intermediating between the call and the calculation should say "you can't ask for that much out".

You're probably right that the price calculation should warn us that the caller is misbehaving, so we can fix any code that might pass a negative in.

Chris-Hibbert commented:
@erights github doesn't seem to let me specify you as a reviewer (presumably because you created it) but I'd like your response anyway.

erights commented on packages/zoe/src/contractSupport/bondingCurves.js:
```diff
@@ -1,7 +1,10 @@
 
-const { add, subtract, multiply, floorDivide } = natSafeMath;
+const { multiply, floorDivide } = natSafeMath;
+const BIG_10000 = BigInt(10000);
```
Just add a TODO comment explaining that we say `BigInt(1000)` rather than `1000n` because some tool somewhere in our toolchain is using some JavaScript parser that does't know how to parse bigint literal syntax. (Even better if you can say what that tool is.)

Chris-Hibbert replied:
Even if `10000n` were legal/worked, I'd still prefer the manifest constant because typos like your `1000n` would be caught more easily.   Maybe BIG_TEN_THOUSAND would be even clearer.

erights replied:
Even then we would rewrite this line as
```js
const BIG_SOMETHING = 1000n;
```

Chris-Hibbert replied:
Got it.  Still a shared manifest constant, but preferentially using the `n` notation.

erights commented on packages/zoe/src/contractSupport/bondingCurves.js:
```diff
@@ -32,12 +35,14 @@ export const getInputPrice = (
-  const denominator = add(multiply(inputReserve, 10000), inputWithFee);
-
-  return floorDivide(numerator, denominator);
+  if (inputReserve <= 0 || outputReserve <= 0 || inputValue < 0) {
+    return 0;
+  }
```
As we discussed, should be error rather than zero. Best is for these errors to come from assert failures.
```suggestion
  assert(inputReserve > 0, details`inputReserve ${inputReserve} must be positive`);
  assert(outputReserve > 0, details`outputReserve ${outputReserve} must be positive`);
  Nat(inputValue);
```

erights commented on packages/zoe/src/contractSupport/bondingCurves.js:
```diff
@@ -65,14 +70,14 @@ export const getOutputPrice = (
-
-  return floorDivide(numerator, denominator);
+  if (inputReserve <= 0 || outputReserve <= 0 || outputReserve <= outputValue) {
+    return 0;
```
Same

erights reviewed (commented):
Other than the zero issue, LGTM

Chris-Hibbert reviewed (commented):
changed `return 0` to `throw Error()`

erights reviewed (commented):
LGTM

As you noticed, I cannot approve. But consider yourself approved. If you can't merge, let me know because I can.

katelynsills reviewed (approved):
LGTM!
